src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs

```
@@ -26,5 +68,61 @@ private static bool GetSwitchValue(string switchName, string envVariable)
         }
+
+        private static bool TryGetStringValue(string switchName, string envVariable, [NotNullWhen(true)] out string? value)
+        {
+            value = AppContext.GetData(switchName) as string;
+            if (string.IsNullOrEmpty(value))
+            {
+                value = Environment.GetEnvironmentVariable(envVariable);
+                if (string.IsNullOrEmpty(value))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static string CreateLibraryName(ReadOnlySpan<char> baseName, ReadOnlySpan<char> suffix, ReadOnlySpan<char> extension, ReadOnlySpan<char> version, bool versionAtEnd = false)
+        {
+            int length = baseName.Length + suffix.Length + version.Length + extension.Length;
+
+            // We validate that suffix and version are not larger than 53 characters.
+            Span<char> result = stackalloc char[length];
+            baseName.CopyTo(result);
+
+            Span<char> secondPart = result.Slice(baseName.Length);
+            suffix.CopyTo(secondPart);
+
+            Span<char> middle = secondPart.Slice(suffix.Length);
+
+            if (!versionAtEnd)
+            {
+                version.CopyTo(middle);
+
+                Span<char> end = middle.Slice(version.Length);
+                extension.CopyTo(end);
+            }
+            else
+            {
+                extension.CopyTo(middle);
+
+                Span<char> end = middle.Slice(extension.Length);
+                version.CopyTo(end);
+            }
+
+            return result.ToString();
+        }
+
+        private static IntPtr LoadLibrary(string library, Assembly assembly, bool failOnLoadFailure)
+        {
+            if (!NativeLibrary.TryLoad(library, assembly, DllImportSearchPath.ApplicationDirectory, out IntPtr lib) && failOnLoadFailure)
+            {
+                Environment.FailFast($"Failed to load app-local ICU: {library}");
```

Should we throw a DllNotFoundException instead of fail fasting?  Again, just want to make sure we've thought through the scenario and what the desired behavior is.  If we've agreed on fail fast here, ok.

---

if the app introduced wrong config switch, I would say fail fast is better here. because will give early clue what can be fixed. I am not sure what will happen if we throw DllNotFoundException in that early execution and at that time we should introduce a fallback path.

---
In reply to: [414603770](https://github.com/dotnet/runtime/pull/35383#discussion_r414603770) [](ancestors = 414603770)

---

Same comment as: https://github.com/dotnet/runtime/pull/35383#discussion_r414746131

cc: @tarekgh ?